pyomo/contrib/doe/DoE_package.py

```diff
@@ -0,0 +1,1812 @@
+
+        if self.scale_nominal_param_value:
+            print('Sensitivity information is scaled by its corresponding parameter nominal value.')
+
+        if (self.scale_constant_value != 1):
+            print('Sensitivity information is scaled by constant ', self.scale_constant_value, ' times itself.')
+
+        if check_mode:
+            # finite or sipopt needs to be chosen
+            if self.mode not in ['simultaneous_finite', 'sequential_finite', 'sequential_sipopt', 'sequential_kaug']:
+                print('Wrong mode. Choose from "simultaneous_finite", "sequential_finite", "0sequential_sipopt", "sequential_kaug"')
+
+        if check_dimension_dv:
+            # input check, only for heatmap
+            # check if there are two design variables and two ranges
+            if not len(self.dv_ranges.keys()) == 2:
+                raise ValueError('Deign variable should be 2')
+
+
+
+    def optimize_doe(self,  design_values, if_optimize=True, obj_opt='det',
+                     scale_nominal_param_value=False, scale_constant_value=1, if_cho=False, L_LB = 1E-10, L_initial=None,
+                     formula='central', step=0.001, check=True):
+        '''
+        Optimize DOE problem with design variables being the decisions.
+        The DOE model is formed invasively and all scenarios are computed simultaneously.
+        The function will first fun a square problem with design variable being fixed at
+        the given initial points, and then unfix the design variable and do the
+        optimization.
+
+        Args:
+            design_values: initial point for optimization, a dict whose keys are design variable names, values are a dict whose keys are time point and values are the design variable value at that time point
+            if_optimize: if True, continue to do optimization. else, just run square problem with given design variable values
+            obj_opt: choose from 'det' and 'trace'
+            scale_nominal_param_value: if scale Jacobian by the corresponding parameter nominal value
+            scale_constant_value: how many order of magnitudes the Jacobian value is magnified by. Use when the Jac or FIM value is too small
+            if_cho: if true, cholesky decomposition is used for Objective function (to optimize determinant).
+                L_LB: if FIM is P.D., the diagonal element should be positive, so we can set a LB like 1E-10
+                L_initial: initialize the L
+            formula: Finite difference formula, choose from 'central', 'forward', 'backward', None
+            step: Finite difference sensitivity perturbation step size, a fraction between [0,1]. default is 0.001
+            check: if True check input toggles consistency to be checked multiple times.
+
+        Returns:
+            analysis_square: result summary of the square problem solved at the initial point
+            analysis_optimize: result summary of the optimization problem solved
+        '''
```

Comment:
I recommend adding an "Actions" or "Steps" section to the doc string. Then listing the steps of this function.
1. Build two-stage stochastic programming optimization model where scenarios correspond to finite difference approximations for the Jacobian of the response variables with respect to calibrated model parameters
2. Fix the experiment design decisions and solve a square (i.e., zero degrees of freedom) instance of the two-stage DOE problem. This step is for initialization.
3. Unfix the experiment design decisions and solve the two-stage DOE problem.